### PR TITLE
Fix periodicity constraints for multiply constrained DoFs

### LINKED ISSUE #5741: Order of boundaries indicators in GridTools::collect_periodic_faces
The following program (modified version of the one reported in https://groups.google.com/forum/#!topic/dealii/k2S8-PKFoik) tests the difference in the computed constraints from `    GridTools::collect_periodic_faces(triangulation, b_id1, b_id2, periodicity_vector);` when exchanging `b_id1` and `b_id2`. For both `ConstraintMatrix` a periodic function is projected on to the ansatz space. Apparently, the results differ which indicates some problem with the computation of the constraints.
```
#include <deal.II/base/quadrature.h>
#include <deal.II/base/function.h>
#include <deal.II/lac/vector.h>
#include <deal.II/lac/constraint_matrix.h>
#include <deal.II/fe/fe_q.h>
#include <deal.II/fe/mapping_q.h>
#include <deal.II/grid/tria.h>
#include <deal.II/grid/grid_generator.h>
#include <deal.II/dofs/dof_tools.h>
#include <deal.II/dofs/dof_handler.h>
#include <deal.II/numerics/data_out.h>
#include <deal.II/numerics/vector_tools.h>
#include <deal.II/grid/grid_refinement.h>
#include <deal.II/grid/grid_tools.h>
#include <deal.II/grid/grid_out.h>
#include <fstream>
#include <iostream>

using namespace dealii;

template <int dim>
ConstraintMatrix make_constraint_matrix(const DoFHandler<dim> &dof_handler, bool minus)
{
  ConstraintMatrix constraints;
  constraints.clear();
  DoFTools::make_hanging_node_constraints(dof_handler, constraints);
  std::vector<GridTools::PeriodicFacePair<typename DoFHandler<dim>::cell_iterator> > periodicity_vectorDof;
  for (int i = 0; i < dim; ++i)
    {
      GridTools::collect_periodic_faces(dof_handler, minus?2*i:2*i+1, minus?2*i+1:2*i, i, periodicity_vectorDof);
    }
  DoFTools::make_periodicity_constraints<DoFHandler<dim> >(periodicity_vectorDof, constraints);
  constraints.close();
  /*std::map<types::global_dof_index, Point<dim> > support_points;
  DoFTools::map_dofs_to_support_points (MappingQ<dim,dim>(1), dof_handler, support_points);
  for (const auto &line: constraints.get_lines())
    for (const auto &entry: line.entries)
      std::cout << "DoF " << line.index << " at " << support_points[line.index]
                << " is constrained to " << " DoF " << entry.first << " at "
                << support_points[entry.first]
                << " with value " << entry.second << std::endl;

  /////write mesh
  std::cout<<"writing mesh .." <<std::endl;
  DataOut<dim> data_out;
  data_out.attach_dof_handler(dof_handler);
  data_out.build_patches ();
  std::ofstream output("mesh.vtu");
  data_out.write_vtu (output);  */
  return constraints;
}

template <int dim>
class PeriodicReference : public Function<dim>
{
public:
  PeriodicReference () : Function<dim>() {}
  virtual double value (const Point<dim>   &p,
                        const unsigned int  component = 0) const override
  {
    return std::sin(p(0)+1.)*std::sin(p(1)+2.)*std::sin(p(2)+3.);
  }
};


int main (int argc, char *argv[])
{
  constexpr int dim = 3;
  const double L=numbers::PI;
  Triangulation<dim> triangulation;
  GridGenerator::hyper_cube (triangulation, -L, L, true);

  std::vector<GridTools::PeriodicFacePair<typename Triangulation<3,3>::cell_iterator> > eriodicity_vector;
  for (int i = 0; i < dim; ++i)
    GridTools::collect_periodic_faces(triangulation, 2*i, 2*i+1, i, periodicity_vector);
  triangulation.add_periodicity(periodicity_vector);

  //two level mesh refinement
  triangulation.refine_global(1);

  typename Triangulation<dim>::active_cell_iterator cellBegin = triangulation.begin_active();
  cellBegin->set_refine_flag();
  triangulation.execute_coarsening_and_refinement();

  /////write mesh
  /*std::cout<<"writing mesh .." <<std::endl;
  DataOut<dim> data_out;
  data_out.attach_dof_handler(dof_handler);
  data_out.build_patches ();
  std::ofstream output("mesh.vtu");
  data_out.write_vtu (output);  */

  FE_Q<dim> FE(1); //linear shape function
  DoFHandler<dim> dof_handler (triangulation);
  dof_handler.distribute_dofs(FE);

  ConstraintMatrix constraints_plus = make_constraint_matrix<dim> (dof_handler, false);
  ConstraintMatrix constraints_minus = make_constraint_matrix<dim> (dof_handler, true);

  PeriodicReference<dim> periodic_function;

  Vector<double> projection_plus(dof_handler.n_dofs());
  Vector<double> projection_minus(dof_handler.n_dofs());

  VectorTools::project (dof_handler, constraints_plus, QGauss<dim>(2), periodic_function, projection_plus);
 VectorTools::project (dof_handler, constraints_minus, QGauss<dim>(2), periodic_function, projection_minus);

  for (unsigned int i=0; i<dof_handler.n_dofs(); ++i)
    std::cout << projection_plus(i)-projection_minus(i) << std::endl;
}
```

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #5741.
Previously, we just omitted an identity constraint if the first of the two degrees of freedom involved was already constrained. This causes problems when asking for periodicity constraints for multiple boundary pairs at the edges between these faces. In particular, we omitted some required constraints.
This PR tries the following: If constraining `dof_1` to `dof_2` doesn't work, because `dof_1` is constrained, and `dof_2` is unconstrained, constraint `dof_2` to `dof_1`.

Furthermore, we detect if adding an identity constraint results in an identity constraint circle. If that is the case the constraint is omitted nonetheless.

### CODE REVIEW

File: source/dofs/dof_tools_constraints.cc [hunk 1]
```diff
@@ -1873,167 +1873,226 @@ namespace DoFTools
...
+                // to check whether this is the case, first see whether there are
+                // any weights other than 0 and 1, then in a first stage make sure
+                // that if so there is only one weight equal to 1
+                //
+                // afterwards do the same for constraints of type dof1=-dof2
+                bool is_identity_constrained = true;
+                const double eps = 1.e-13;
+                for (unsigned int jj=0; jj<dofs_per_face; ++jj)
+                  if (((std::abs(transformation(i,jj)) < eps) ||
+                       (std::abs(transformation(i,jj)-1) < eps)) == false)
```
Comment (tjhei) [inline source/dofs/dof_tools_constraints.cc]: can you do ``1.``? The ``== false`` is also more complicated than I would have written it.

File: source/dofs/dof_tools_constraints.cc [hunk 2]
```diff
@@ -1873,167 +1873,226 @@ namespace DoFTools
...
+                        }
+                  }
+
+                bool is_inverse_constrained = !is_identity_constrained;
+                unsigned int inverse_constraint_target = numbers::invalid_unsigned_int;
+                if (is_inverse_constrained)
                   for (unsigned int jj=0; jj<dofs_per_face; ++jj)
                     if (((std::abs(transformation(i,jj)) < eps) ||
-                         (std::abs(transformation(i,jj)-1) < eps)) == false)
+                         (std::abs(transformation(i,jj)+1) < eps)) == false)
```
Comment (tjhei) [inline source/dofs/dof_tools_constraints.cc]: 1.

File: source/dofs/dof_tools_constraints.cc [hunk 3]
```diff
@@ -1873,167 +1873,226 @@ namespace DoFTools
...
+                          constraint_matrix.add_line(dofs_2[i]);
+                          constraint_set = true;
+                        }
+                  }
+
+                if (!constraint_set)
+                  {
+                    // now treat constraints, either as an equality constraint or
+                    // as a sequence of constraints
+                    if (is_identity_constrained == true || is_inverse_constrained == true)
```
Comment (tjhei) [inline source/dofs/dof_tools_constraints.cc]: I would remove the ``==true``.

**Comment by masterleinad** [comment]
It's probably better to look at the diff without whitespace changes, i.e. https://github.com/dealii/dealii/pull/5779/files?w=1. Anyway, I will fix up your findings as well.

**Comment by masterleinad** [comment]
@tjhei I addressed your comments.

**Comment by tjhei** [comment]
Looks good, but please fix indentation.

**Comment by masterleinad** [comment]
@tjhei Here you go!

**Comment by jppelteret** [comment]
@masterleinad I know that this PR has been sitting here for a couple of days, but lets give it one more day for others to comment on. If there is no further discussion by tomorrow then I think that you should feel free to merge it yourself then.

**Pull request merged**